persistence-api/src/main/java/io/stargate/db/datastore/PersistenceBackedDataStore.java

```diff
@@ -1,81 +1,178 @@
+        executeParameters.protocolVersion() == parameters().protocolVersion(),
+        "Invalid execution parameters: cannot modify the protocol version for execution "
+            + "(the DataStore version %s != %s, the execution parameters version).",
+        parameters().protocolVersion(),
+        executeParameters.protocolVersion());
   }
 
   @Override
-  public CompletableFuture<PreparedStatement> prepare(String queryString) {
+  public <B extends BoundQuery> CompletableFuture<Query<B>> prepare(Query<B> query) {
     return connection
-        .prepare(queryString, parameters)
-        .thenApply(
-            prepared ->
-                new PersistenceBackedPreparedStatement(
-                    connection, parameters, new PreparedInfo(prepared), queryString));
+        .prepare(query.queryStringForPreparation(), parameters())
+        .thenApply(prepared -> query.withPreparedId(prepared.statementId));
   }
 
   @Override
-  public CompletableFuture<ResultSet> batch(List<String> queries) {
+  public CompletableFuture<ResultSet> execute(
+      BoundQuery query, UnaryOperator<Parameters> parametersModifier) {
     long queryStartNanos = System.nanoTime();
-    List<Statement> persistenceStatements =
-        queries.stream().map(SimpleStatement::new).collect(Collectors.toList());
-    return batch(persistenceStatements, BatchType.LOGGED, parameters, queryStartNanos);
+    Parameters executeParameters = parametersModifier.apply(parameters());
+    validateExecuteParameters(executeParameters);
+
+    CompletableFuture<ResultSet> future = new CompletableFuture<>();
+    if (query.source().query().preparedId().isPresent()) {
+      executeWithRetry(query, executeParameters, queryStartNanos, future);
+    } else if (options.alwaysPrepareQueries()) {
+      prepareAndRetry(query.source(), executeParameters, queryStartNanos, future);
+    } else {
+      doExecute(query, executeParameters, queryStartNanos, future, future::completeExceptionally);
+    }
+    return future;
+  }
+
+  private void executeWithRetry(
+      BoundQuery query,
+      Parameters executeParameters,
+      long queryStartNanos,
+      CompletableFuture<ResultSet> future) {
+    doExecute(
+        query,
+        executeParameters,
+        queryStartNanos,
+        future,
+        ex -> {
+          if (ex instanceof PreparedQueryNotFoundException) {
+            // This could happen due to a schema change between the statement preparation and now,
+            // as some schema change can invalidate preparation.
+            logger.debug(
+                "Prepared statement (id={}) was invalid when executed. This can happen due to a "
+                    + "conflicting schema change. Will re-prepare and retry.",
+                ((PreparedQueryNotFoundException) ex).id);
+            prepareAndRetry(query.source(), executeParameters, queryStartNanos, future);
+          } else {
+            future.completeExceptionally(ex);
+          }
+        });
```

Comment:
Is there a possibility of an infinite retry chain here?.. perhaps a very remote possibility if, for example, `PreparedQueryNotFoundException` never resolves into a more severe exception during retries. WDYT?

Reply:
I honestly don't think there is a risk and I don't personally think it's worth adding a stop condition.

I mean, a stop condition would have to be a fixed number of retries. But you're juggling the risk of C* having a serious bug (and knowing what the preparation code in C* is, that feels very remote indeed) against the risk of failing a query that would have worked with just one more retry. And it's all super remote chances, so I'm not sure picking one versus the other matters.

Besides, as this is a recursive calls, we actually cannot technically have an "infinite" retry chain. Java will throw a stack overflow exception soon enough. I don't think that's a worth exception to be thrown that what we would do if we added a manual stop condition; in both case, this almost surely indicate a bug that needs fixing.

But anyway, that is _my_ reasoning of why I didn't bother adding a stop condition (in short: it adds a little bit of logic complexity, for no good reason imo), but it's opinionated. If you guys are uncomfortable with that opinion, I'll comply and add the condition.

Reply:
Makes sense. I somehow thought the call was returning without recursion :facepalm: